File: src/domain/recursion/execution-graph-sync.ts
import type { RecursiveModelConfig, RecursivePromptMetadata } from "../types.js";
import { estimateModelCalls, estimateToolRounds, remainingModelCalls } from "./budget-guard.js";

/** Node type from the live execution graph snapshot. */
export type LiveExecutionNode = NonNullable<
  RecursivePromptMetadata["executionGraph"]
>["nodes"][number];

/** Edge type from the live execution graph snapshot. */
export type LiveExecutionEdge = NonNullable<
  RecursivePromptMetadata["executionGraph"]
>["edges"][number];

/** Build `executionGraph` + live `budget` fields from orchestrator maps/counters. */
export function buildLiveExecutionMetadata(input: {
  executionNodes: Map<string, LiveExecutionNode>;
  executionEdges: LiveExecutionEdge[];
  modelCalls: number;
  maxModelCalls: number;
  toolCallsLength: number;
  toolRoundLimit: number;
  config?: RecursiveModelConfig | undefined;
}): Pick<RecursivePromptMetadata, "executionGraph" | "budget"> {
  const config = input.config;
  return {
    executionGraph: {
      nodes: [...input.executionNodes.values()],
      edges: [...input.executionEdges],
    },
    budget: {
      estimatedModelCalls: estimateModelCalls(config, input.modelCalls),
      estimatedToolRounds: estimateToolRounds(input.toolRoundLimit, config),
      modelCallsUsed: input.modelCalls,
      modelCallsRemaining: remainingModelCalls(input.modelCalls, input.maxModelCalls),
      toolCallsUsed: input.toolCallsLength,
    },
  };
}

File: crates/rlm-core/src/domain/recursion/execution_graph_sync.rs
use std::collections::HashMap;

use crate::domain::recursion::{estimate_model_calls, estimate_tool_rounds, remaining_model_calls};
use crate::domain::types::{
    ExecutionBudget, ExecutionGraph, ExecutionGraphEdge, ExecutionGraphNode, RecursiveModelConfig,
};

pub fn build_live_execution_metadata(
    execution_nodes: &HashMap<String, ExecutionGraphNode>,
    execution_edges: &[ExecutionGraphEdge],
    model_calls: u32,
    max_model_calls: u32,
    tool_calls_length: u32,
    tool_round_limit: u32,
    config: Option<&RecursiveModelConfig>,
) -> (ExecutionGraph, ExecutionBudget) {
    let graph = ExecutionGraph {
        nodes: execution_nodes.values().cloned().collect(),
        edges: execution_edges.to_vec(),
        viewport: None,
    };
    let budget = ExecutionBudget {
        estimated_model_calls: estimate_model_calls(config, model_calls),
        estimated_tool_rounds: estimate_tool_rounds(tool_round_limit, config),
        model_calls_used: model_calls,
        model_calls_remaining: remaining_model_calls(model_calls, max_model_calls),
        tool_calls_used: tool_calls_length,
    };
    (graph, budget)
}

#[cfg(test)]
#[path = "../../../tests/domain/recursion/execution_graph_sync.rs"]
mod execution_graph_sync_tests;
